sdk/java/pulumi/src/main/java/com/pulumi/resources/CustomTimeouts.java

```diff
@@ -92,6 +92,78 @@ public static String golangString(Optional<Duration> duration) {
         return duration.get().toNanos() + "ns";
     }
 
+    /**
+     * Given a string representing a duration, such as "300ms", "-1.5h" or "2h45m", parse it into
+     * a {@link Duration}.
+     *
+     * @param timeout the string to parse
+     * @return the parsed duration, or null if the input string was empty or null
+     */
+    public static Duration parseTimeoutString(String timeout) {
+        if (timeout == null || timeout.isEmpty()) {
+            return null;
+        }
+
+        // A duration string is a possibly signed sequence of decimal numbers, each with optional
+        // fraction and a unit suffix, such as "300ms", "-1.5h" or "2h45m". Valid time units are "ns",
+        // "us" (or "µs"), "ms", "s", "m", "h".
+
+        String s = timeout;
+        boolean neg = false;
+        if (s.charAt(0) == '-' || s.charAt(0) == '+') {
+            neg = s.charAt(0) == '-';
+            s = s.substring(1);
+        }
+        if (s.equals("0")) {
+            return Duration.ZERO;
+        }
+        if (s.isEmpty()) {
+            throw new IllegalArgumentException("invalid duration " + timeout);
+        }
+
+        Duration duration = Duration.ZERO;
+        while (!s.isEmpty()) {
+            // find the next timeunit
+            int i = 0;
+            while (i < s.length() && (Character.isDigit(s.charAt(i)) || s.charAt(i) == '.')) {
+                i++;
+            }
+            // parse the number
+            double v = Double.parseDouble(s.substring(0, i));
+            // parse the unit
+            s = s.substring(i);
+            if (s.isEmpty()) {
+                throw new IllegalArgumentException("missing unit in duration " + timeout);
+            }
+
+            if (s.startsWith("ns")) {
+                duration = duration.plusNanos((long)v);
+                s = s.substring(2);
+            } else if (s.startsWith("µs") || s.startsWith("us")) {
+                duration = duration.plusNanos((long)(v * 1000));
+                s = s.substring(2);
+            } else if (s.startsWith("ms")) {
+                duration = duration.plusNanos((long)(v * 1_000_000));
+                s = s.substring(2);
+            } else if (s.startsWith("s")) {
+                duration = duration.plusNanos((long)(v * 1_000_000_000));
```

Comment:
```suggestion
                duration = duration.plusSeconds((long)v);
```

Would be easier to read I think.

Reply:
Yeah, but we lose precision if we round v to a long here (and plusSeconds accepts a long).

Reply:
Ah because the value can be with commas, I missed that.  Yeah `plusNanos` makes sense them.